### PR TITLE
Track which user created resources

### LINKED ISSUE #13306: Proposal: Record which user created an object
Currently no creator information is recorded at the time an object is created. The creator may be:
- an end user calling the api (authenticated via the Authenticator interfaces)
- a controller calling the api (authenticated via Authenticator)
- a core controller locally invoking the storage interfaces (bypasses authentication)
- someone directly editing etcd (repair process or object created in the past).

It is a common requirement to record audit information with objects that allow UIs and administrators to determine the actor involved in a transaction. The information may be recorded in a side channel (an audit log), but end users may also wish to understand who created a particular object.

Use cases:
- Allow clients to determine which user made a change
- Allow an administrator to reconstruct an audit record for who created which resources were created in the system

In the first two scenarios above, the user is clearly identified (`GetUserName()` on the user resolved by the authenticator, which rest.Storage has access to).  In the third, this is effectively what we call a "system" user, and could be assigned an arbitrary username that does not collide with real user names, or left blank.  In the last case there is probably a user name, but no mechanism whereby it is provided.
#### Inferential attribution
- Actions may not be undertaken by a user, but on behalf of a user
  - the replication controller process creates pods on behalf of the user who created a replication controller, but if another user edits the replication controller, both users may have a causal relationship to the pods created
  - a virtual resource may run as elevated privileges against another api and allow a user to create a resource they normally don't have permission to create - both the end user and the api process have a causal relationship to the underlying resource
  - a higher privilege user may request to run as another user via **impersonation** - the actor is the higher privilege user, but they are explicitly wishing to be recorded as acting as the impersonated user
- Actions may be separated in time and space from their cause
  - a user may delete a node, which results in a pod being deleted, which results in a pod being created by a replication controller

Recommendation: we should scope the attribution of creation to the caller (direct creator), and ensure that an external "audit log" could be constructed that can infer chain of attribution for complex cases.
#### Creating child resources

In general, it may be desirable for high level constructs (deployments, scheduled jobs) that create lower level resources (replication controllers, jobs) to propagate creation attribution. However, that raises a security risk that the creating component can lie about the source of a creation. Since an audit trail can be reconstructed between parent and subordinate resources by observing API changes, it is not necessary to propagate that info.
#### Referencing a user

We currently define a `user name` as a string that represents a unique user. That value is expected to be stable over the time periods the admin desires (only admin action would result in a user name changing). Under the covers, the user name may refer to many different types of users - the [authentication design](https://github.com/kubernetes/kubernetes/blob/master/docs/design/access.md) identifies two:
- User
  - ServiceAccount - a namespaced resource that represents a unique automated identity
    - In the future, this account may be backed by an external user
  - ExternalUser - via an external system
    - May be a human or an external service account

In addition, we have an implicit type:
- SystemUser - created implicitly in the infrastructure to represent distinct process identity prior to bootstrapping of service accounts or external users
  - Which user represents the code in the master that creates the default service account?

The `Authenticator` interface requires all of the above users to be mapped into a single user name string before filtering through the system. This string is the `userName` and must be unique across all sources of authentication.

It is assumed that the audit system and other client systems understand the mapping between the various authentication sources and `userName`.  These mappings may differ between different installations.  It should be possible for a client or third party to provide a `userName` and retrieve an object (or objects?) representing the source of the user name. 

As an example, Kube must overlay the service account user identifier onto the external authenticator source. Today, that is enforced via a unique prefix `system:serviceaccount:` that is reserved. In the future, we may map `ServiceAccount` to an external user, in which case the unique prefix may not be needed. OpenShift, which needs to manage users internally as well as externally, uses a further system of subdivision of namespace to assign a `userName` via indirection.
### Proposed change

When objects are created through rest.Storage, record the identifying user info onto the object. Add `creationUserName string` to `ObjectMeta` and assign it the `GetUserName()` value provided by the Authenticator. This value may not be changed, but may be unset if authentication is disabled.

Implications for clients:
- Clients will want to show "pretty" names to end users (`userName` may be email, but may also be a unique identifier or LDAP DN or short name). Clients that wish to do that will need a way to map `userName` to a useful value efficiently and in bulk (it's assumed that the "pretty name" for a user changes at very infrequent intervals)
- Clients may be tempted to assume the `userName` can be parsed. This limits the admin's ability to change this value. Instruct clients not to assume the `userName` value will be human readable or that it follows a parseable pattern.

Other implications:
- Clients should not be allowed to set the `creationUserName`. 
- Administrators may wish to migrate old data to assign a `creationUserName`.

### PR DESCRIPTION
**Pull request by smarterclayton**
Adds `creationUserName` which records the creating user in the resource
for use in audit scenarios. Set by the system automatically to the user
name returned by the authentication plugin. Clients must assume that the
name value may be machine-readable and therefore is not guaranteed to be
a pretty name.

Fixes #13306

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/13925)

### CODE REVIEW

**Comment by smarterclayton** [comment]
@kubernetes/kube-api - the proposal didn't get many comments but this adds a new metadata field to track created users.

**Comment by derekwaynecarr** [comment]
I assume ```CreationUserName``` is preferred over ```Creator``` because it aligns with ```CreationTimestamp```?

File: pkg/api/meta.go [hunk 1]
```diff
@@ -25,6 +25,9 @@ import (
 // FillObjectMetaSystemFields populates fields that are managed by the system on ObjectMeta.
 func FillObjectMetaSystemFields(ctx Context, meta *ObjectMeta) {
 	meta.CreationTimestamp = util.Now()
+	if user, ok := UserFrom(ctx); ok {
```
Comment (derekwaynecarr) [inline pkg/api/meta.go]: When would there never be a user?

Comment (smarterclayton) [inline pkg/api/meta.go]: Yes, I don't want to create new names even if the english is slightly
weird.  We're fairly good at "<concept><type>" in our API - "creation" is
the concept, "type" is "UserName", "Timestamp", "PeriodInSeconds", "Name",
"Ref"

On Mon, Sep 14, 2015 at 1:41 PM, Derek Carr <notifications@github.com>
wrote:

> In pkg/api/meta.go
> <https://github.com/kubernetes/kubernetes/pull/13925#discussion_r39422744>
> :
>
> > @@ -25,6 +25,9 @@ import (
> >  // FillObjectMetaSystemFields populates fields that are managed by the system on ObjectMeta.
> >  func FillObjectMetaSystemFields(ctx Context, meta *ObjectMeta) {
> >  	meta.CreationTimestamp = util.Now()
> > +	if user, ok := UserFrom(ctx); ok {
>
> When would there never be a user?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13925/files#r39422744>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

Comment (smarterclayton) [inline pkg/api/meta.go]: There could not be a user if it's an object created before this change went
in, created by a core system process (like endpoints and namespaces created
by the master on startup), or created by direct edit in etcd that does not
require authentication.  Or a cluster that doesn't require authentication
or doesn't pass that info to Kube.

On Mon, Sep 14, 2015 at 2:08 PM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> Yes, I don't want to create new names even if the english is slightly
> weird.  We're fairly good at "<concept><type>" in our API - "creation" is
> the concept, "type" is "UserName", "Timestamp", "PeriodInSeconds", "Name",
> "Ref"
>
> On Mon, Sep 14, 2015 at 1:41 PM, Derek Carr <notifications@github.com>
> wrote:
>
>> In pkg/api/meta.go
>> <https://github.com/kubernetes/kubernetes/pull/13925#discussion_r39422744>
>> :
>>
>> > @@ -25,6 +25,9 @@ import (
>> >  // FillObjectMetaSystemFields populates fields that are managed by the system on ObjectMeta.
>> >  func FillObjectMetaSystemFields(ctx Context, meta *ObjectMeta) {
>> >  	meta.CreationTimestamp = util.Now()
>> > +	if user, ok := UserFrom(ctx); ok {
>>
>> When would there never be a user?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/13925/files#r39422744>.
>>
>
>
>
> --
> Clayton Coleman | Lead Engineer, OpenShift
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by smarterclayton** [comment]
ok to test

**Comment by smarterclayton** [comment]
@bgrant0607 let me know if the naming makes sense here

**Comment by bgrant0607** [comment]
I understand that you want a field that can't be set/modified by the client, as opposed to an annotation. I'm fine with that.

CreationUserName seems fine with respect to the field naming conventions. 

I also like that CreationUserName makes it clear that it's a user name and not, for instance, a controller name.

What I don't know is what our plans for User concepts/APIs are. Is "UserName" consistent with our current thinking in that direction? @erictune? @liggitt? @derekwaynecarr?

File: pkg/api/validation/validation.go [hunk 2]
```diff
@@ -273,6 +273,12 @@ func ValidateObjectMetaUpdate(new, old *api.ObjectMeta) errs.ValidationErrorList
 	} else {
 		new.CreationTimestamp = old.CreationTimestamp
 	}
+	// ignore changes to user name
+	if len(old.CreationUserName) == 0 {
+		old.CreationUserName = new.CreationUserName
```
Comment (deads2k) [inline pkg/api/validation/validation.go]: If it was empty before, it seems weird to allow it to be updated later.

**Comment by smarterclayton** [comment]
On Sep 16, 2015, at 1:24 AM, Brian Grant <notifications@github.com> wrote:

I understand that you want a field that can't be set/modified by the
client, as opposed to an annotation. I'm fine with that.

CreationUserName seems fine with respect to the field naming conventions.

I also like that CreationUserName makes it clear that it's a user name and
not, for instance, a controller name.

What I don't know is what our plans for User concepts/APIs are. Is
"UserName" consistent with our current thinking in that direction? @erictune
<https://github.com/erictune>? @liggitt <https://github.com/liggitt>?
@derekwaynecarr <https://github.com/derekwaynecarr>?


As defined in our authentication patterns we settled on UserName as the
unique in space name for a user.  We don't have a way to enforce uniqueness
for time and space today within Kube (in Openshift we do) because
underlying auth providers are responsible for that (as defined by their
system).  An admin is free to let their username not be stable if they
like.  I would maybe argue in the future to expand this to include userUID,
but not required for step one.


—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/13925#issuecomment-140629217>
.

File: pkg/api/validation/validation.go (hunk 2 above)
Comment (smarterclayton) [inline pkg/api/validation/validation.go]: It was to allow an automated process to update it for legacy data.  So
someone upgrading from Kube 1.0 would be able to be assigned the value.  I
can disable the ability to set it post creation, which means we have no API
path to ever set that value and would have to write directly to etcd.
Allowing it to be set from empty to something post create has potential for
abuse.

On Sep 16, 2015, at 9:47 AM, David Eads <notifications@github.com> wrote:

In pkg/api/validation/validation.go
<https://github.com/kubernetes/kubernetes/pull/13925#discussion_r39630766>:

> @@ -273,6 +273,12 @@ func ValidateObjectMetaUpdate(new, old *api.ObjectMeta) errs.ValidationErrorList
>  	} else {
>  		new.CreationTimestamp = old.CreationTimestamp
>  	}
> +	// ignore changes to user name
> +	if len(old.CreationUserName) == 0 {
> +		old.CreationUserName = new.CreationUserName

If it was empty before, it seems weird to allow it to be updated later.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/13925/files#r39630766>.

Comment (liggitt) [inline pkg/api/validation/validation.go]: +1 for making it immutable by default

**Comment by liggitt** [comment]
CreationUserName seems ok. The user.Info interface has both Name() and UID(). Do we want to preserve both (that gets us point in time uniqueness as well).

**Comment by smarterclayton** [comment]
I wanted to make the minimal change here to start down the path.  If
everyone is ok with UID(), we can add it.  On the other hand, we can easily
add it later.

On Wed, Sep 16, 2015 at 12:57 PM, Jordan Liggitt <notifications@github.com>
wrote:

> CreationUserName seems ok. The user.Info interface has both Name() and
> UID(). Do we want to preserve both (that gets us point in time uniqueness
> as well).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13925#issuecomment-140803389>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by smarterclayton** [comment]
Rebased.  This is as minimal a change as I can make that provides real value.  Do we have consensus on `creationUserName` as the name?  I think we can add `creationUserUID` in the future.  Other thoughts?

**Comment by deads2k** [comment]
code lgtm.

I like username as a starting point.

**Comment by liggitt** [comment]
If we think we'd add uid at some point, I have a vague desire for an object reference instead to keep the two fields paired, but I'd be ok with deferring that (especially since I don't know what I would set the kind to)

**Comment by smarterclayton** [comment]
I had the same reasoning.  Ok, so any final comments from anyone?  This
would then just be API review on whether Brian shares the similar concern
about future move of this to an object ref (where we don't have a Kind yet
for user so it wouldn't be exactly an object ref).

On Fri, Sep 18, 2015 at 11:40 AM, Jordan Liggitt <notifications@github.com>
wrote:

> If we think we'd add uid at some point, I have a vague desire for an
> object reference instead to keep the two fields paired, but I'd be ok with
> deferring that (especially since I don't know what I would set the kind to)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13925#issuecomment-141485684>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by erictune** [comment]
I'm fine with starting with just UserName and adding UserID later.  

I have questions about the proposal which I will direct to #13306

**Comment by erictune** [comment]
For the audit log use case, I think logging apiserver requests is more useful than this.

For the "who made this thing so that I can get help", we should use the created-by annotation. 
Kubectl could optionally (by default?) set this annotation whenever creating an object.

**Comment by smarterclayton** [comment]
Ok, I think I see the perspective (a different place that where I came
from).

Kubectl does not know this value (in an authoritative way), so the field
would not be practically machine readable for any serious system of record
(the user identifier the authentication system sees is != what the
identifier a client may know).  So the value would not be of practical use
to a client concerned with reporting it since its use is optional.

The place where this proposal comes from is driven by historical choices in
enterprise software - at the point of record of a transaction, record the
identity of the actor that was authorized by the system in the core
resource.  In some systems this is ownership - in others it is audit data.

This fields has first order value to human users of the API (and second
order value to machine users like audit and activity recorders).  An admin
or user viewing the system having a verifiable identifier associated with
objects that can be used to track and comprehend the various participants
in a system.  I don't think any optional, user settable field achieves
that.  Having a secondary audit record is valuable, but relying on that to
report this data is much harder (extra calls and Apis that don't exist).

So this to me sits in between "strong audit" and "weak attribution" as
"weak audit, strong attribution" and is valuable to consumers of the API
today at minimal cost.


On Sep 18, 2015, at 5:27 PM, Eric Tune <notifications@github.com> wrote:

For the audit log use case, I think logging apiserver requests is more
useful than this.

For the "who made this thing so that I can get help", we should use the
created-by annotation.
Kubectl could optionally (by default?) set this annotation whenever
creating an object.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/13925#issuecomment-141573286>
.

**Comment by smarterclayton** [comment]
Thoughts?  Should we take this to slack or a hangout?

**Comment by smarterclayton** [comment]
@erictune regarding my last comment I think we have two options - expose a `metadata` field that is enforced by the api server as write-once and handled under the covers, or use an admission controller to enforce them strongly (not allow modifications once set, always use the user from the api.Context).  I was concerned with exposing annotations users can't modify, but I don't believe in practice admins / users expect these values to be user settable.

Do we want to have admission controllers that impose/enforce semantics around annotations?   If so, I can make that change, and we're likely to a) always run with it and b) silently drop or edit users who don't mutate them.  If not, or if we can agree these are strongly typed fields that are a core part of the schema, then I would want to proceed with adding them to metadata.

**Comment by erictune** [comment]
I feel moderately strongly that we should not add creation or modification user names to our metadata.

As an audit function, it is a weak choice, for the reasons you mentioned and others.  As attribution, I don't think it is all that useful for a couple of reasons:

1. pods are created by controllers, not end users.  joining this data at controller-creation time makes the controllers more complex and not worth it.  not propagating this data will make the field useful.
1. creation user not useful for when resources are updated by different user than creator.

**Comment by erictune** [comment]
What if, in the apiserver, after an object has been successfully created or updated, but before we have completed handling the HTTP request (and so still have the context), we generated an event about that object that includes the username, e.g.: "Pod foo created by user clayton", "Pod bar updated by user eric", "Pod bar deleted by user daniel".

Some things I like about that approach:

- it gives a nicer timeline of what happened to an object, including creation, modification, and deletion.  
- no new API needed to view info, since we already have a way to view events
- different clusters can have different limits on who can see what events, thus allowing control over who can see this info
- if appropriate permissions set, events cannot be deleted by non-super-users, so it makes a stronger audit log
- no new admission controller needed.

**Comment by erictune** [comment]
@smarterclayton

**Comment by erictune** [comment]
@saad-ali because he :heart:  events.

**Comment by smarterclayton** [comment]
Downside is that is a lot more events.  Also, today at least events are owned by a namespace so they can be deleted when the namespace is deleted.

**Comment by erictune** [comment]
It may be worth comparing with other IaaS/PaaS APIs which are mature enough
to have well-developed audit capabilities.  I picked cloud foundry and
openstack for comparison.


-  cloud foundry apps have the following metadata: uid,  object name,
"created_at", "updated_at" (source:
https://apidocs.cloudfoundry.org/226/apps_(experimental)/starting_an_app.html
)
-  cloud foundry API audit trail is accessed with `cf events` command.
(source: https://docs.cloudfoundry.org/concepts/security.html#logging)
- openstack metadata includes a request ID, creation time, object uid,
object name (source:
http://developer.openstack.org/api-ref-compute-v2.1.html)
- openstack auditing is an optional feature (source:
http://docs.openstack.org/developer/keystonemiddleware/audit.html). It uses
a log format which records events (source:
https://www.dmtf.org/sites/default/files/standards/documents/DSP2038_1.1.0.pdf
)

**Comment by smarterclayton** [comment]
The salesforce metadata is another good metric of what and end state
metadata audit looks like:

https://help.salesforce.com/apex/HTViewHelpDoc?id=field_audit_trail.htm&language=en_US

Tracks creator and updater.

I think event tracking makes a lot of sense for approaching audit, but does
require more thought in how events are implemented and handled.  For
instance, having events warehoused, survive namespace deletion, and
protected/exposed safely to end users.

**Comment by cameronbrunner** [comment]
Just curious how this request is progressing.  Specifically the created-by-user is of great interest but not for an audit perspective but more as a pod attribute.  As discussed in #19079 (and sorry for those who have read both of these comments!) I am interested in this as we are working on integrating our scheduler (Grid Engine) with kubernetes and from our past experience user ownership of compute tasks is important when organizations implement resource allocation policies.  It would be useful for us to be able to quota against users, favor pods created by one user over another, and group pods of certain users under hierarchical organizational groups.  With our scheduler we could do just about all of this now in kubernetes if there was a user tag on resources.

**Comment by erictune** [comment]
@cameronbrunner Resources are often created by controllers in kubernetes, not end-users.  Before we can have this feature in a useful way, we would need to implement acting-as #16933.

**Comment by cameronbrunner** [comment]
@erictune  Thanks for the reference to #16933 thats definitely covering the ownership issue with resource creation inside of controllers in general.  I was thinking I was going to have to have to do something on my own here so its good to see that people have already been thinking about this:)

**Comment by fejta** [comment]
This PR has had no meaningful activity for multiple months. If it is still valid please rebase, push a new commit and reopen the PR. Thanks!

**Comment by davidopp** [comment]
Why did we drop this? It seems useful.

**Comment by aanm** [comment]
@erictune @smarterclayton Since https://github.com/kubernetes/kubernetes/pull/23549 was merged, what is the current progress with this PR? Let me know if I can help with anything.

**Comment by bgrant0607** [comment]
cc @mikedanese

**Comment by erictune** [comment]
https://github.com/kubernetes/kubernetes/pull/27087 will add a detailed audit log.

We should see how many use cases that satisfies (I know it does not satisfy @cameronbrunner use case, but want to see how many other need this after we have an audit log).

Also, submitting this before we have implemented controllers that use "acting-as" will mean that we have to do a breaking change once we add "acting-as" support.  So, we should do that first, and then reevaluate this.

**Comment by erictune** [comment]
@aanm It would be helpful if you could describe in detail how you would use this field.

**Comment by aanm** [comment]
@erictune It would help to allow or deny network traffic from/to a particular user.

**Comment by tgraf** [comment]
@erictune The use case is derived from the Net-SIG NetworkPolicy decision which now allows ingress rules based on PodSelector. We would like to allow for the creation of rules based on who has created/is running a particular pod.

**Comment by erictune** [comment]
Why not do this based on the service account of the pod?  This allows two good things:

- a single human can create different pods with different network policies
- multiple humans can collaborate in the management of pods for an important service.

**Comment by soltysh** [comment]
@sttts fyi

**Pull request closed without merging**